### ISSUE TITLE
Jobs getting locked and stuck after 4 or 5 successive runs.

### ISSUE THREAD

**Issue by sampathBlam** [agenda/agenda#483]
We have used agenda in our node.js server for certain background polling operations. We have defined three jobs in agenda. 

`agenda.define('pollDatasetStatus', function (job, done) {
    // Job to poll the status of some dataset in a thirdparty server.
})`

`agenda.define('pollExecutionStatus', function (job, done) {
    // Job to poll the status of the execution of a dataset in the thirdparty server.    
})`

`agenda.define('sendEmailToUsers', function (job, done) {
    // Job to send email
})`

We start agenda using `agenda.start()` in the `ready` event listener of agenda. 
We create and schedule each job instance in the following manner:

`var pollDatasetJob = agenda.create('pollDatasetStatus',  { // some job params });
pollDatasetJob.repeatEvery('1 minute');
pollDatasetJob.save();`

`var pollExecutionJob = agenda.create('pollExecutionStatus',  { // some job params });
pollExecutionJob .repeatEvery('1 minute');
pollExecutionJob.save();`

`var sendEmailJob = agenda.create('sendEmailToUsers',  { // some job params });
sendEmailJob.repeatEvery('1 minute');
sendEmailJob.save();`

Inside each job, using the job attributes, we maintain a counter using which we restrict the maximum number of times each job gets executed after which we remove the job from the queue using `job.remove()`. 

In case of 'pollDatasetStatus' and 'pollExecutionStatus', based on the desired state of dataset and execution, we either call `done()` to complete the job or call `done()` and `job.remove()` to complete and remove the job from queue. 

For eg.

In case of 'pollDatasetStatus' : 
if dataset status is desired:
   // we complete and remove the job.
   `done(); job.remove()`
else 
   // we complete the job.
   `done()`

The flow of our jobs would be : pollDatasetStatus -> pollExecutionStatus -> sendEmailToUsers.

Multiple users can trigger a pollDatasetStatus agenda job creation via our express endpoint, and hence, multiple instances of each job can run concurrently. 

**The issue is that, sometimes, when a job is created in the agenda database, it has a non-null value in the lockedAt column. Once this happens, all the successive jobs that are created have non-null value at the lockedAt column and no job gets executed.** 

```
{ 
    "_id" : ObjectId("5976e30ccbaa420c24ab6a8b"),
    "name" : "pollDatasetStatus",
    "data" : {
       // some private data
        "attemptCount" : 0
    },
    "type" : "normal",
    "priority" : 0,
    "nextRunAt" : ISODate("2017-07-25T06:19:56.549Z"),
    "repeatInterval" : "1 minute",
    "repeatTimezone" : null,
    "lastModifiedBy" : null,
    "lockedAt" : ISODate("2017-07-25T06:19:56.550Z")
}

{
    "_id" : ObjectId("5976e352cbaa420c24ab6a8c"),
    "name" : "pollDatasetStatus",
    "data" : {
        // some private data
        "attemptCount" : 0
    },
    "type" : "normal",
    "priority" : 0,
    "nextRunAt" : ISODate("2017-07-25T06:21:06.418Z"),
    "repeatInterval" : "1 minute",
    "repeatTimezone" : null,
    "lastModifiedBy" : null,
    "lockedAt" : ISODate("2017-07-25T06:21:06.420Z")
}
{
    "_id" : ObjectId("5976e363cbaa420c24ab6a8d"),
    "name" : "pollDatasetStatus",
    "data" : {
        // some private data
        "attemptCount" : 0
    },
    "type" : "normal",
    "priority" : 0,
    "nextRunAt" : ISODate("2017-07-25T06:21:23.536Z"),
    "repeatInterval" : "1 minute",
    "repeatTimezone" : null,
    "lastModifiedBy" : null,
    "lockedAt" : ISODate("2017-07-25T06:21:23.537Z")
}

```
Once the node.js server is restarted, all the agenda jobs start executing one by one and their lockedAt column becomes null. 

Another weird thing we noted is that after 10 minutes, which is the defaultLockLifetime, the lockedAt value of all the locked jobs seem to get updated. But they never get executed.

Versions: 
Agenda : 0.9.1
Node : 6.9.5
Mongoose: 4.10.4

**Comment by sampathBlam** [comment]
We have found that the above issue is due to the following sequence of actions.

- Agenda seems to maintain **counters** for running and locked jobs in memory for each job definition. 

- As and when a job is locked, it increments the **locked** counter and when a job is run, it increments the **running** counter for the particular job definition. 

- And when a job is done ( **done()** is called on the job) , the counters are decremented.

- In our code, for the above three job definitions, we call job.remove() without calling done() on the job when we want to finish up a job in certain cases.

- As we just called job.remove() without calling done(), the counters were never decremented back. And soon they reached the default concurrency limit of 5. 

- From that point, all the jobs get locked and never get run as the **running** counter is always equal to the concurrency limit. 

This seems to be the reason for our issue. I should accept that this issue is more due to our ignorance of how agenda's locking mechanism works than anything to with agenda itself. Hence closing this issue.
